README.md
# GraphQL Custom Types
[![npm version](https://badge.fury.io/js/graphql-custom-types.svg)](https://badge.fury.io/js/graphql-custom-types) [![Build Status](https://secure.travis-ci.org/stylesuxx/graphql-custom-types.png?branch=master)](https://travis-ci.org/stylesuxx/graphql-custom-types) [![Dependency Status](https://david-dm.org/stylesuxx/graphql-custom-types.svg)](https://david-dm.org/stylesuxx/graphql-custom-types) [![devDependencies Status](https://david-dm.org/stylesuxx/graphql-custom-types/dev-status.svg)](https://david-dm.org/stylesuxx/graphql-custom-types?type=dev) [![peerDependencies Status](https://david-dm.org/stylesuxx/graphql-custom-types/peer-status.svg)](https://david-dm.org/stylesuxx/graphql-custom-types?type=peer)

> This is a collection of custom GraphQL types that I tend to reuse quite often so I packed them into a module.

## Available Types
Let me give you an overview of the available types. If you need more detail about how to use them, check *schema.js* in the tests folder.

### Scalar
The primitive types, aka everything that may be represented as a string. The ones with parameters you need to instantiate with *new* and pass according parameters, the others may be used as are.

* GraphQLEmail
* GraphQLURL
* GraphQLDateTime
* GraphQLLimitedString(min, max, alphabet)
* GraphQLPassword(min, max, alphabet, complexity)
* GraphQLUUID

*complexity* default options:
```JavaScript
{
  alphaNumeric: false,
  mixedCase: false,
  specialChars: false
}
```

## Installation
Most likely you already will have it, but do not forget to also install *graphql*, since it is required as peer dependency:
```Bash
npm install graphql graphql-custom-types --save
```

## Usage
```JavaScript
import {
  GraphQLEmail,
  GraphQLURL,
  GraphQLDateTime,
  GraphQLLimitedString,
  GraphQLPassword,
  GraphQLUUID
} from 'graphql-custom-types';
```

And use it in your Schema as you would use any other type.

## Development
Contributions are very welcome, please feel free to submit a type. If you do so make sure there are test cases in place.

### Testing
The test suite may be invoked by running:
```Bash
npm run test
```


CHANGELOG.md
# Changelog
## [1.7.0] - 2022-09-14
- Updated graphql peer dependency to ^16

## [1.6.0] - 2020-07-03
- Updated graphql peer dependency
- Dropped support for NodeJs < 6.9
- Improved .gitignore

## [1.5.1] - 2019-07-10
### Fixed
- TS Password definition

## [1.5.0] - 2018-11-22
- Updated graphql peer dependency

## [1.4.0] - 2018-05-20
- Updated graphql peer dependency

## [1.3.0] - 2018-01-31
- Updated graphql peer dependency
- Updated dev dependencies
- Adjusted test cases - graphql ^0.12.0 is not throwing errors when validation fails, instead it is populating an error property in the result object. Testing will fail with older graphql versions.

## [1.2.0] - 2017-10-02
- Updated peer dependency

## [1.1.0] - 2017-06-28
- Added typescript definitions
- Added editorconfig

## [1.0.0] - 2017-06-19
- Updated peer dependency

## [0.7.3] - 2017-05-05
- Updated peer dependency

## [0.7.2] - 2016-11-15
- Updated dev dependencies

## [0.7.1] - 2016-11-14
- Updated graphql dependencies

## [0.7.0] - 2016-09-20
### Improved
- Added badges for dev and peer dependencies

### Updated dependencies
- graphql
- eslint

### Fixed
- Properly import Factory

## [0.6.0] - 2016-06-09
### Improved
- graphql is now a peerDependecy

### Fixed
- Date validation

## [0.5.1] - 2016-06-08
### Updated dependencies
- graphql to ^0.6.0

## [0.5.0] - 2016-06-08
### Imporoved
- Properly parse values

## [0.4.0] - 2016-05-10
### Added
- Export factory and GraphQLCustomScalarType
- Test against nodeJS >= 4.4 && <= 6.1

### Removed
- Support for iojs
- Support for nodeJS < 4.4

## [0.2.0] - 2015-12-12
### Added
- GraphQLDateTime

## [0.1.1] - 2015-12-11
### Improved
- Packaging for NPM

## [0.1.0] - 2015-12-11
### Added
- Custom scalar factory
- RegExp type factory
- GraphQLURL
- GraphQLLimitedString
- GraphQLPassword
- CHANGELOG

### Improved
- Test cases
- README
- Improved NPM config, exlude files that are not needed to use the library

### Deprecated
- Dropped support for NodeJs < 0.11.0

## [0.0.0] - 2015-12-9
### Added
- GraphQLEmail


tests/schema.js
import {
  graphql,
  GraphQLSchema,
  GraphQLObjectType,
  GraphQLString
} from 'graphql';
import {
  GraphQLEmail,
  GraphQLURL,
  GraphQLLimitedString,
  GraphQLPassword,
  GraphQLDateTime,
  GraphQLUUID
} from '../lib';

export const schema = new GraphQLSchema({
  query: new GraphQLObjectType({
    name: 'RootQueryType',
    fields: {
      email: {
        type: GraphQLString,
        args: {
          item: { type: GraphQLEmail }
        },
        resolve: (root, {item}) => {
          return item;
        }
      },
      url: {
        type: GraphQLString,
        args: {
          item: { type: GraphQLURL }
        },
        resolve: (root, {item}) => {
          return item;
        }
      },
      limitedStringDefault: {
        type: GraphQLString,
        args: {
          item: { type: new GraphQLLimitedString() }
        },
        resolve: (root, {item}) => {
          return item;
        }
      },
      limitedStringMinMax: {
        type: GraphQLString,
        args: {
          item: { type: new GraphQLLimitedString(3, 10) }
        },
        resolve: (root, {item}) => {
          return item;
        }
      },
      limitedStringAlphabet: {
        type: GraphQLString,
        args: {
          item: { type: new GraphQLLimitedString(3, 10, 'abc123') }
        },
        resolve: (root, {item}) => {
          return item;
        }
      },
      password: {
        type: GraphQLString,
        args: {
          item: { type: new GraphQLPassword(null, null, null, { alphaNumeric: true }) }
        },
        resolve: (root, {item}) => {
          return item;
        }
      },
      passwordMixedCase: {
        type: GraphQLString,
        args: {
          item: { type: new GraphQLPassword(null, null, null, { mixedCase: true }) }
        },
        resolve: (root, {item}) => {
          return item;
        }
      },
      passwordSpecialChars: {
        type: GraphQLString,
        args: {
          item: { type: new GraphQLPassword(null, null, null, { specialChars: true }) }
        },
        resolve: (root, {item}) => {
          return item;
        }
      },
      passwordAll: {
        type: GraphQLString,
        args: {
          item: { type: new GraphQLPassword(3, 6, 'abcABC123!"§', { specialChars: true, mixedCase: true, alphaNumeric: true }) }
        },
        resolve: (root, {item}) => {
          return item;
        }
      },
      date: {
        type: GraphQLString,
        args: {
          item: { type: GraphQLDateTime }
        },
        resolve: (root, {item}) => {
          return item;
        }
      },
      uuid: {
        type: GraphQLString,
        args: {
          item: { type: GraphQLUUID }
        },
        resolve: (root, {item}) => {
          return item;
        }
      },
    }
  })
});


index.d.ts
// tslint:disable max-classes-per-file
import {GraphQLScalarType} from "graphql";

declare namespace graphqlCustomTypes {

  type IASTParser<TResult = any> = (ast: {kind: string, value: string}) => TResult;

  interface IFactoryOptions {
    name: string;
    regex: RegExp;
    description: string;
    error?: string;
  }

  class Factory {
    public getRegexScalar(options: IFactoryOptions): GraphQLCustomScalarType;

    public getCustomScalar(name: string, description: string, parser: IASTParser): GraphQLCustomScalarType;
  }

  class GraphQLCustomScalarType extends GraphQLScalarType {
    constructor(name: string, description: string, parser: IASTParser);
  }

  const GraphQLEmail: GraphQLCustomScalarType;

  const GraphQLURL: GraphQLCustomScalarType;

  const GraphQLDateTime: GraphQLCustomScalarType;

  const GraphQLUUID: GraphQLCustomScalarType;

  class GraphQLPassword extends GraphQLCustomScalarType {
    constructor(min?: number, max?: number, alphabet?: string, complexity?: {
      alphaNumeric?: boolean,
      mixedCase?: boolean,
      specialChars?: boolean,
    });
  }

  class GraphQLLimitedString extends GraphQLCustomScalarType {
    constructor(min?: number, max?: number, alphabet?: boolean)
  }
}

export = graphqlCustomTypes;


src/scalar.js
import {
  GraphQLString,
  GraphQLScalarType
} from 'graphql';

import { GraphQLError } from 'graphql/error';
import { Kind } from 'graphql/language';

export const GraphQLEmail = new GraphQLScalarType({
  name: 'Email',
  description:'The Email scalar type represents E-Mail addresses compliant to RFC 822.',
  serialize: value => {
    return value;
  },
  parseValue: value => {
    return value;
  },
  parseLiteral: ast => {
    if (ast.kind !== Kind.STRING) {
      throw new GraphQLError('Query error: Can only parse strings got a: ' + ast.kind, [ast]);
    }

    var re = /^(([^<>()[\]\.,;:\s@\"]+(\.[^<>()[\]\.,;:\s@\"]+)*)|(\".+\"))@(([^<>()[\]\.,;:\s@\"]+\.)+[^<>()[\]\.,;:\s@\"]{2,})$/i;
    if(!re.test(ast.value)) {
      throw new GraphQLError('Query error: Not a valid Email address', [ast]);
    }

    return ast.value;
  }
});


src/factory.js
import {
  GraphQLString,
  GraphQLScalarType
} from 'graphql';
import { GraphQLError } from 'graphql/error';
import { Kind } from 'graphql/language';
import { GraphQLCustomScalarType } from './types';

export class Factory {
  getRegexScalar(options) {
    const error = options.error || 'Query error: ' + options.name;

    const parser = function(ast) {
      if (ast.kind !== Kind.STRING) {
        throw new GraphQLError('Query error: Can only parse strings got a: ' + ast.kind, [ast]);
      }

      var re = options.regex;
      if(!re.test(ast.value)) {
        throw new GraphQLError(error, [ast]);
      }

      return ast.value;
    };

    return this.getCustomScalar(options.name, options.description, parser);
  }

  getCustomScalar(name, description, parser) {
    return new GraphQLCustomScalarType(name, description, parser);
  }
}


src/scalars.js
import { GraphQLError } from 'graphql/error';
import { Kind } from 'graphql/language';
import { Factory } from './factory';
import { GraphQLCustomScalarType } from './types';

const factory = new Factory();

export const GraphQLEmail = factory.getRegexScalar({
  name: 'Email',
  regex: /^(([^<>()[\]\.,;:\s@\"]+(\.[^<>()[\]\.,;:\s@\"]+)*)|(\".+\"))@(([^<>()[\]\.,;:\s@\"]+\.)+[^<>()[\]\.,;:\s@\"]{2,})$/i,
  description: 'The Email scalar type represents E-Mail addresses compliant to RFC 822.',
  error: 'Query error: Not a valid Email address'
});

export const GraphQLURL = factory.getRegexScalar({
    name: 'URL',
    // RegExp taken from https://gist.github.com/dperini/729294
    regex: new RegExp('^(?:(?:https?|ftp)://)(?:\\S+(?::\\S*)?@)?(?:(?!(?:10|127)(?:\\.\\d{1,3}){3})(?!(?:169\\.254|192\\.168)(?:\\.\\d{1,3}){2})(?!172\\.(?:1[6-9]|2\\d|3[0-1])(?:\\.\\d{1,3}){2})(?:[1-9]\\d?|1\\d\\d|2[01]\\d|22[0-3])(?:\\.(?:1?\\d{1,2}|2[0-4]\\d|25[0-5])){2}(?:\\.(?:[1-9]\\d?|1\\d\\d|2[0-4]\\d|25[0-4]))|(?:(?:[a-z\\u00a1-\\uffff0-9]-*)*[a-z\\u00a1-\\uffff0-9]+)(?:\\.(?:[a-z\\u00a1-\\uffff0-9]-*)*[a-z\\u00a1-\\uffff0-9]+)*(?:\\.(?:[a-z\\u00a1-\\uffff]{2,}))\\.?)(?::\\d{2,5})?(?:[/?#]\\S*)?$', 'i'),
    description: 'The URL scalar type represents URL addresses.',
    error: 'Query error: Not a valid URL'
});

export const GraphQLUUID = factory.getRegexScalar({
    name: 'UUID',
    // https://github.com/chriso/validator.js/blob/master/src/lib/isUUID.js#L7
    regex: new RegExp('^[0-9A-F]{8}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{12}$', 'i'),
    description: 'The UUID scalar type represents a UUID.',
    error: 'Query error: Not a valid UUID'
});

const stringValidator = function(ast) {
  if (ast.kind !== Kind.STRING) {
    throw new GraphQLError('Query error: Can only parse strings got a: ' + ast.kind, [ast]);
  }
};

const lengthValidator = function(ast, min, max) {
  if(ast.value.length < min) {
    throw new GraphQLError('Query error: String not long enough', [ast]);
  }

  if(max && ast.value.length > max) {
    throw new GraphQLError('Query error: String too long', [ast]);
  }
};

const alphabetValidator = function(ast, alphabet) {
  for(var char of ast.value) {
    if(alphabet.indexOf(char) < 0) {
      throw new GraphQLError('Query error: Invalid character found', [ast]);
    }
  }
};

const complexityValidator = function(ast, options) {
  const complexity = options || {};
  const alhpaNumericRe = /^(?=.*[0-9])(?=.*[a-zA-Z])(.+)$/;
  const mixedCaseRe = /^(?=.*[a-z])(?=.*[A-Z])(.+)$/;
  const specialCharsRe = /^(?=.*[^a-zA-Z0-9])(.+)$/;

  if(complexity.alphaNumeric && !alhpaNumericRe.test(ast.value)) {
    throw new GraphQLError('Query error: String must contain at least one number and one letter', [ast]);
  }

  if(complexity.mixedCase && !mixedCaseRe.test(ast.value)) {
    throw new GraphQLError('Query error: String must contain at least one uper and one lower case letter', [ast]);
  }

  if(complexity.specialChars && !specialCharsRe.test(ast.value)) {
    throw new GraphQLError('Query error: String must contain at least one special character', [ast]);
  }
};

var limitedStringCounter = 0;
export class GraphQLLimitedString extends GraphQLCustomScalarType {
  constructor(min = 1, max, alphabet) {
    const suffix = (limitedStringCounter++ > 0) ? limitedStringCounter : '';
    const name = 'LimitedString' + suffix;
    var description = 'A limited string.';
    if(max) description += ' Has to be between ' + min + ' and ' + max + ' characters long.';
    else description += ' Has to be at least ' + min + ' characters long.';
    if(alphabet) description += ' May only contain the following characters: ' + alphabet;

    const validator = function(ast) {
      stringValidator(ast);
      lengthValidator(ast, min, max);

      if(alphabet) alphabetValidator(ast, alphabet);

      return ast.value;
    }

    super(name, description, validator);
  }
};

var passwordCounter = 0;
export class GraphQLPassword extends GraphQLCustomScalarType {
  constructor(min = 1, max, alphabet, complexity) {
    const suffix = (passwordCounter++ > 0) ? passwordCounter : '';
    const name = 'Password' + suffix;
    var description = 'A password string.';
    if(max) description += ' Has to be between ' + min + ' and ' + max + ' characters long.';
    else description += ' Has to be at least ' + min + ' characters long.';
    if(alphabet) description += ' May only contain the following characters: ' + alphabet;
    if(complexity) {
      if(complexity.alphaNumeric) description += ' Has to be alpha numeric.';
      if(complexity.mixedCase) description += ' Has to be mixed case.';
      if(complexity.specialChars) description += ' Has to contain special characters.';
    }

    const validator = function(ast) {
      stringValidator(ast);
      lengthValidator(ast, min, max);

      if(alphabet) alphabetValidator(ast, alphabet);
      if(complexity) complexityValidator(ast, complexity);

      return ast.value;
    }

    super(name, description, validator);
  }
};

export const GraphQLDateTime = factory.getCustomScalar(
  'DateTime',
  'The DateTime scalar type represents date time strings complying to ISO-8601.',
  function(ast) {
    stringValidator(ast);
    if(isNaN(Date.parse(ast.value))) {
      throw new GraphQLError('Query error: String is not a valid date time string', [ast]);
    }

    return ast.value;
  }
);
